### ISSUE TITLE
The --bootstrap-patches option only changes the KubeadmConfig API but not Configmap in workload cluster

### ISSUE THREAD

**Issue by williamsandrew** [vmware/cluster-api-upgrade-tool#144]
When supplying a set of JSON patches via the `--bootstrap-patches` option the KubeadmConfig API object in the management cluster is updated successfully. However, the Kubeadm configmap in the workload cluster is not changed which means said patches are effectively ignored. The only change that happens to the Kubeadm configmap is the `version` field which comes from this code: https://github.com/vmware/cluster-api-upgrade-tool/blob/0288bc748c450b487caceae130a9822df5d47a46/pkg/upgrade/control_plane.go#L204-L206.

To replicate, run the upgrade tool with a `--bootstrap-patches` option similar to this:

```
--bootstrap-patches=[{"op":"replace","path":"/spec/clusterConfiguration/dns/imageRepository","value":"example.com"},{"op":"replace","path":"/spec/clusterConfiguration/dns/imageTag","value":"some-tag"}]
```

Now review the KubeadmConfig API object for the upgraded machine as well as the Kubeadm Configmap in the workload cluster. You'll notice that only the former changed the DNS options.

**Comment by ncdc** [comment]
@fabriziopandini what is the best way to handle this, where the user wants to change the coredns/etcd images used for new control plane machines when upgrading? Is it appropriate to modify the kubeadm configmap?

**Comment by fabriziopandini** [comment]
@ncdc 
TL;DR; immutable upgrade means `kubeadm join`, and `kubeadm join` was not designed to handle "change the cluster"; in the long term, the solution for change the cluster will be the kubeadm-operator.

In the meantime, you can leverage on the fact that `kubeadm join` uses the kubeadm configmap as a source of truth for all the cluster-wide settings (except node-specific settings).

So, by changing the kubeadm configmap before running the upgrade tool, you will get the new etcd pods generated by `kubeadm join` use the new etcd image.

But unfortunately this address your problem only partially because `kubeadm join` does note re-creates the core DNS addon, so this requires a different approach, like e.g use phases

**Comment by ncdc** [comment]
@williamsandrew sounds like you may need to do something custom for CoreDNS changes.
